policy/predicate/signature.go

```diff
@@ -0,0 +1,150 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/palantir/policy-bot/policy/common"
+	"github.com/palantir/policy-bot/pull"
+)
+
+type HasValidSignatures bool
+
+var _ Predicate = HasValidSignatures(false)
+
+func (pred HasValidSignatures) Evaluate(ctx context.Context, prctx pull.Context) (bool, string, error) {
+	commits, err := prctx.Commits()
+	if err != nil {
+		return false, "", errors.Wrap(err, "failed to get commits")
+	}
+
+	if !pred {
+		return true, "", nil
+	}
+
+	for _, c := range commits {
+		valid, desc := hasValidSignature(ctx, c)
+		if !valid {
+			return false, desc, nil
+		}
+	}
+
+	return true, "", nil
+}
+
+func (pred HasValidSignatures) Trigger() common.Trigger {
+	return common.TriggerCommit
+}
+
+type HasValidSignaturesBy struct {
+	common.Actors `yaml:",inline"`
+}
+
+var _ Predicate = &HasValidSignaturesBy{}
+
+func (pred *HasValidSignaturesBy) Evaluate(ctx context.Context, prctx pull.Context) (bool, string, error) {
+	commits, err := prctx.Commits()
+	if err != nil {
+		return false, "", errors.Wrap(err, "failed to get commits")
+	}
+
+	signers := make(map[string]struct{})
+
+	for _, c := range commits {
+		valid, desc := hasValidSignature(ctx, c)
+		if !valid {
+			return false, desc, nil
+		}
+		signers[c.Signature.GetSigner()] = struct{}{}
+	}
+
+	for signer := range signers {
+		member, err := pred.IsActor(ctx, prctx, signer)
+		if err != nil {
+			return false, "", err
+		}
+		if !member {
+			return false, fmt.Sprintf("Contributor %q does not meet the required membership conditions for signing", signer), nil
+		}
+	}
+
+	return true, "", nil
+}
+
+func (pred *HasValidSignaturesBy) Trigger() common.Trigger {
+	return common.TriggerCommit
+}
+
+type HasValidSignaturesByKeys struct {
+	KeyIDs []string `yaml:"key_ids"`
+}
+
+var _ Predicate = &HasValidSignaturesByKeys{}
+
+func (pred *HasValidSignaturesByKeys) Evaluate(ctx context.Context, prctx pull.Context) (bool, string, error) {
+	commits, err := prctx.Commits()
+	if err != nil {
+		return false, "", errors.Wrap(err, "failed to get commits")
+	}
+
+	keys := make(map[string]struct{})
+
+	for _, c := range commits {
+		valid, desc := hasValidSignature(ctx, c)
+		if !valid {
+			return false, desc, nil
+		}
+		// Only GPG signatures are valid for this predicate
+		switch sig := c.Signature.(type) {
+		case *pull.GPGSignature:
+			keys[sig.GetKeyID()] = struct{}{}
+		default:
+			return false, fmt.Sprintf("Commit %.10s signature is not a GPG signature", c.SHA), nil
+		}
+	}
+
+	for key := range keys {
+		isValidKey := false
+		for _, acceptedKey := range pred.KeyIDs {
+			if key == acceptedKey {
+				isValidKey = true
```

Comment:
```suggestion
				isValidKey = true
				break
```

Although I could also see converting `pred.KeyIDs` to a map or moving this inner loop check to a helper.

Reply:
👍